### ISSUE TITLE
Update iocp reactor to manage correctly the new _stopped flag

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3200]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#3200|
|Type|release blocker: regression|
|Created|2008-04-22 13:10:42Z|
|Branch|https://github.com/twisted/twisted/tree/iocp-fix-3200|

Since r23345, `ReactorBase` now uses a new _stopped flag to correctly manage its running state. We should update the iocp reactor to use it, because it's not working correctly anymore.

The correct workflow might have been to revert the changeset, but we concluded that it could be bypassed here.

<details><summary>Searchable metadata</summary>

```
trac-id__3200 3200
type__release_blocker__regression release blocker: regression
reporter__therve therve
priority__high high
milestone__Twisted_8_1 Twisted-8.1
branch__branches_iocp_fix_3200 branches/iocp-fix-3200
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1208869842000000 1208869842000000
changetime__1208889401000000 1208889401000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve commented|
|-|-|

(In [23378]) Branching to 'iocp-fix-3200'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @exarkun|
|-|-|

Please have a look.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

This looks pretty good.

Can you add a bit to the `_SignalReactorMixin` docstring?  The mixin expects some methods to be implemented on the class mixing it in - eg, sigInt, sometimes sigBreak, callLater, mainLoop.  It also pretty much requires that it be mixed in with ReactorBase, since it calls up to a method of that class.  Maybe it would make more sense if there to be _SignalReactorBase subclassing ReactorBase and subclassed by everyone else?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [23390]) Add some doc, remove some more duplication.

Refs [#3200](https://github.com/twisted/twisted/issues/3200)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

Okay, I updated documentation a bit. It's not awesome yet, so if you have some more idea don't hesitate. I didn't change the class hierarchy for now, because I think this is in flux so we just have to wait for something a little bit better. Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Looks good to me, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [23392]) Merge iocp-fix-3200

Author: therve
Reviewer: exarkun
Fixes [#3200](https://github.com/twisted/twisted/issues/3200)

Fix the iocpreactor management of the new _stopped attribute introduced
in r23345. It removes a fair amount of duplication between iocpreactor and
posixbase in the process.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
